src/data/polynomial/derivative.lean

```diff
@@ -148,6 +128,69 @@ lemma iterate_derivative_C_mul (a : R) (p : R[X]) (k : ℕ) :
   derivative^[k] (C a * p) = C a * (derivative^[k] p) :=
 by convert iterate_derivative_smul a p k; apply C_mul'
 
+theorem of_mem_support_derivative {p : R[X]} {n : ℕ} (h : n ∈ p.derivative.support) :
+  n + 1 ∈ p.support :=
+mem_support_iff.2 $ λ (h1 : p.coeff (n+1) = 0), mem_support_iff.1 h $
+show p.derivative.coeff n = 0, by rw [coeff_derivative, h1, zero_mul]
+
+theorem degree_derivative_lt {p : R[X]} (hp : p ≠ 0) : p.derivative.degree < p.degree :=
+(finset.sup_lt_iff $ bot_lt_iff_ne_bot.2 $ mt degree_eq_bot.1 hp).2 $ λ n hp, lt_of_lt_of_le
+(with_bot.some_lt_some.2 n.lt_succ_self) $ finset.le_sup $ of_mem_support_derivative hp
+
+theorem degree_derivative_le {p : R[X]} : p.derivative.degree ≤ p.degree :=
+if H : p = 0 then le_of_eq $ by rw [H, derivative_zero] else (degree_derivative_lt H).le
+
+theorem nat_degree_derivative_lt {p : R[X]} (hp : p.derivative ≠ 0) :
+  p.derivative.nat_degree < p.nat_degree :=
+have hp1 : p ≠ 0, from λ h, hp $ by rw [h, derivative_zero],
+with_bot.some_lt_some.1 $
+begin
+  rw [nat_degree, option.get_or_else_of_ne_none $ mt degree_eq_bot.1 hp, nat_degree,
+    option.get_or_else_of_ne_none $ mt degree_eq_bot.1 hp1],
+  exact degree_derivative_lt hp1
+end
+
+theorem nat_degree_derivative_le (p : R[X]) : p.derivative.nat_degree ≤ p.nat_degree :=
+let _ := classical.prop_decidable in by exactI
+if h : p.derivative = 0 then h.symm ▸ zero_le _ else (nat_degree_derivative_lt h).le
+
+@[simp] lemma derivative_cast_nat {n : ℕ} : derivative (n : R[X]) = 0 :=
+begin
+  rw ← map_nat_cast C n,
+  exact derivative_C,
+end
+
+lemma iterate_derivative_eq_zero {p : R[X]} {x : ℕ} (hx : p.nat_degree < x) :
+  polynomial.derivative^[x] p = 0 :=
+begin
+  induction h : p.nat_degree using nat.strong_induction_on with _ ih generalizing p x,
+  subst h,
+  obtain ⟨t, rfl⟩ := nat.exists_eq_succ_of_ne_zero (pos_of_gt hx).ne',
+  rw [function.iterate_succ_apply],
+  by_cases hp : p.derivative = 0,
+  { rw [hp, iterate_derivative_zero] },
+  have := nat_degree_derivative_lt hp,
+  exact ih _ this (this.trans_le $ nat.le_of_lt_succ hx) rfl
+end
+
+theorem nat_degree_eq_zero_of_derivative_eq_zero [no_zero_divisors R] [char_zero R] {f : R[X]}
```

Comment:
Perhaps `∃ x, p = C x` would be a more useful way to state the conclusion of this?

Reply:
this theorem already exists, so if we change it probably in some follow-up PR. there's a lemma that is an iff between this and that, though